Add `sv_vstring_get` API around `PERL_MAGIC_vstring`

leonerd opened the pull request:
This PR adds a new API function, `sv_vstring_get` for interacting with what is currently implemented as `PERL_MAGIC_vstring`. By using this API callers no longer need to know about details of how that magic is implemented (or that it even is magic in the first place).

This is for similar reasons to the API added by #22971, as it gives future flexibility on how this behaviour is actually implemented.

I believe that the `Storable` module is supposed to be dual-life, so I added back-compat wrappers in `Storable.xs` to make it continue to work on older Perls before this API was added.

tonycoz commented on sv.c:
```diff
@@ -17878,6 +17876,47 @@ Perl_sv_regex_global_pos_clear(pTHX_ SV *sv)
+{
+    PERL_ARGS_ASSERT_SV_VSTRING_SET;
+
+    sv_magic(sv, NULL, PERL_MAGIC_vstring, pv, len);
```
What is meant to happen if sv already has a vstring set?

sv_magic() will silently refuse to add the same magic again.

For normal assignment sv_force_normal_flags() removes the vstring magic.

leonerd replied:
Hmmm yes that's a fun question. The existing use-cases where this API was called in core or `Storable.xs` would know that the SV didn't already have the magic, but as a more general API it might not know that. Perhaps `set` should update any existing magic to change the stored values? This is going to mean a `Safefree()` and `savepvn()` call. Hrm.

leonerd replied:
Hrm, thinking further: I can add some code for that case, but there's currently no code that would exercise it to test it. I guess I can add some in XS-APItest or somesuch.

leonerd replied:
Now with behaviour updated and some tests added.

tonycoz commented on dist/Storable/Storable.xs:
```diff
@@ -296,6 +296,28 @@ typedef STRLEN ntag_t;
 #define VSTRING_CROAK() CROAK(("Cannot retrieve vstring in this perl"))
 #endif
 
+#ifndef sv_vstring_get
+#define sv_vstring_get(sv,pvp,lenp) S_sv_vstring_get(aTHX_ sv,pvp,lenp)
+static bool S_sv_vstring_get(pTHX_ SV *sv, const char **pvp, STRLEN *lenp)
+{
+  MAGIC *mg;
+  if(!SvMAGICAL(sv) || !(mg = mg_find(sv, PERL_MAGIC_vstring)))
+    return FALSE;
+
+  *pvp  = mg->mg_ptr;
+  *lenp = mg->mg_len;
+  return TRUE;
+}
+#endif
+
+#ifndef sv_vstring_set
+#define sv_vstring_set(sv,pv,len) \
+  STMT_START {                                             \
+    sv_magic((sv), NULL, PERL_MAGIC_vstring, (pv), (len)); \
+    SvRMAGICAL_on(sv);                                     \
+  } STMT_END
+#endif
+
```
This belongs in Devel::PPPort

Leont replied:
I agree

leonerd replied:
That sounds like an excellent idea that I fully support someone else doing.

Really - I have no idea how to edit or maintain PPPort. The entire process seems utterly opaque to me. If someone else wants to copypaste it over there then I can update this PR to match.

Leont replied:
There is a more general problem that almost no-one knows how to do that. I suspect that the real solution is that several of us learn to do so because otherwise this keeps being a recurring problem.

khwilliamson replied:
I can help getting this into D:P

khwilliamson replied:
So it's fine with me to merge this, and we will release a new D:P later

leonerd replied:
It sounds like overall the process for adding this to `D::P` is nontrivial enough, that I'd suggest it might be easier to first merge this PR as-is, and then as a separate change, move these support helpers out to there. That way it doesn't hold up this particular change and the work depending on it.

bulk88 commented on sv.c:
```diff
@@ -17878,6 +17876,47 @@ Perl_sv_regex_global_pos_clear(pTHX_ SV *sv)
+*/
+
+void
+Perl_sv_vstring_set(pTHX_ SV * const sv, const char *pv, STRLEN const len)
```
Why isn't Perl_sv_vstring_set() sanitizing the user's input args `const char *pv`, `STRLEN len`? Why or why not, are both good answers.

leonerd replied:
I don't follow the question at all.

bulk88 commented:
What happens if arg ` SV* dest_sv` passed by caller into `sv_vstring_set()` is already `IOK_on`/`NOK_on`/`POK_on`/`ROK_on`/`SvIsCOW_on`?

What if PP-level `SvIVX()` vs `SvNVX()` vs `SvPVX()` vs `sv_vstring_set(sv,pv,len)` are 4 different random [i.e. uninit mem] PP `$scalar` values? What should `sv_vstring_set()` do? something? nothing? any answer is a good answer

bulk88 commented on sv.c:
```diff
@@ -17878,6 +17876,47 @@ Perl_sv_regex_global_pos_clear(pTHX_ SV *sv)
+
+If the given SV has vstring magic, stores the string pointer and length into
+the variables addressed by C<pvp> and C<lenp>, and returns true.  If not,
+returns false.
```
I suggest a sentence saying,
```
stores the string pointer and length into
the variables addressed by C<pvp> and C<lenp>. The lifespan
of string pointer is until the next FREETMPS;.
```
or
```
The lifespan of string pointer is until next LEAVE;.
```
or
```
The lifespan of string pointer is of undefined storage and from
undefined allocator.   Memcpy() it ASAP if necessary.  It is not documented
 if string pointer is backed by a Newx()/Safesysmalloc() block,
somewhere into the middle of a Newx() block, or is a SVPV COW,
or is a shared string table HEK or is a temp ptr owned by MORTAL
or SAVESTACK or is a process global ptr to a C const "" string literal.
```

Version # 3  documents lifetime as TBD/opaque/UB.

leonerd replied:
None of those apply; but I've now added some words to point out the string buffer is owned by the SV itself (or at least, the magic annotation on it)

bulk88 commented on dist/Storable/Storable.xs:
```diff
@@ -2597,18 +2620,16 @@ static int store_scalar(pTHX_ stcxt_t *cxt, SV *sv)
 
 #ifdef SvVOK
-        if (SvMAGICAL(sv) && (mg = mg_find(sv, 'V'))) {
+        if (sv_vstring_get(sv, &vstr_pv, &vstr_len)) {
```
If I were a serializing XS mod, and performance is important, `sv_vstring_get` needs to be `#define SvVSTRING(sv, pv, n)` with a fast shortcut on `SvMAGICAL(sv) &&` before executing "heavy" extern exported  PLT/GOT C symbol `Perl_sv_vstring_get()`.

bulk88 commented on sv.c:
```diff
@@ -17878,6 +17876,47 @@ Perl_sv_regex_global_pos_clear(pTHX_ SV *sv)
+
+bool
+Perl_sv_vstring_get(pTHX_ SV * const sv, const char **pvp, STRLEN *lenp)
+{
```
Why is retval of `Perl_sv_vstring_get` of type `bool` and not of type `STRLEN`?

What does `if(SvVOK(sv) && SvVCUR(sv) == 0) { NOOP; }` mean on a PP/XS/C level?

why not move arg 3 `STRLEN *lenp` to `retval`?

 Code like `if(!len) { die("bad input"); }` has universal meaning.

```
if(1) {
        sv_setpvs(TARG, "0 but true");
        XPUSHTARG;
        RETURN;
    }
```
is very rare in Perl/all SW dev.

leonerd replied:
> Why is retval of Perl_sv_vstring_get of type bool and not of type STRLEN?

It matches the similar API shape I added with `sv_regex_global_pos_get()`.

mohawk2 replied:
I like the idea of more use of `retval`s where no ambiguity is possible. Should `sv_regex_global_pos_get()` also return its `STRLEN`?

bulk88 replied:
> sv_regex_global_pos_get()

```
bool
Perl_sv_regex_global_pos_get(pTHX_ SV *sv, STRLEN *posp, U32 flags)
{
```
HELL YEAH.
![getpos](https://github.com/user-attachments/assets/2e3e2d36-3bee-4020-9e0a-3c41edea17c2)

Although `int Perl_magic_getpos(interpreter *my_perl, sv *sv, magic *mg)` is too tiny and too simple to have this defect. MSVC for `len` as in `pv = SvPV(sv, len);` will redundantly re-read `len` over and over from C stack memory, because C stack auto var `len` ESCAPED!!!! any and all C functions in Perl's virtual address space can unpredictably rewrite the value inside C stack auto `len` stored in real C stack memory (not a register).

So it would be beneficial on all ABI/CPU arcs, to return `STRLEN *posp` either as the 1 and only retval, and eliminate the `bool`. or say retval `-1` is failure, everything else is success (retval `0` length is success ??? idk this area of API).

Backup option, return by copy a C struct with 2 members, `bool` and `STRLEN`. Now caller func frame has ABI proof nothing can ever rewrite `STRLEN`, and `STRLEN` can be upgraded to a non-vol register for the rest of this func body.

Also lets us be reasonable, unless `STRLEN` type is transporting a ptr size integer holding a CSRNG value, we know the sign bit aka 0x8000_0000  2GB or 0x8000_0000_0000_0000 aka 9,223,372,036,854,775,808 9 quintillionB will always be zero in the retval, and not the user's string length. After that assumption, STRLEN can transport a 1 bit bool, and a 31 bit or 63 bit integer, inside 1 CPU register, inside 1 C retval.

tonycoz commented on sv.c:
```diff
@@ -17878,6 +17876,64 @@ Perl_sv_regex_global_pos_clear(pTHX_ SV *sv)
+}
+
+/*
+=for apidoc sv_vstring_set
+
+Applies vstring magic to the given SV, storing the string given by C<pv> for
+C<len> bytes into it.  If the SV already had vstring magic, the previous
+stored string is freed and replaced.
+
+=cut
```
This should probably point at upg_version() and new_version(), which is what you want if you're not using a pre-parsed version.

tonycoz replied:
Thinking about this...

Storable uses this mechanism to restore the magic to match an SV it was originally provided, eg. from a `vX.Y`:
```
$ perl -MDevel::Peek -e 'Dump(v1.1.2)'
SV = PVMG(0x556625c6bd00) at 0x556625c29fe0
  REFCNT = 1
  FLAGS = (RMG,POK,IsCOW,READONLY,PROTECT,pPOK)
  IV = 0
  NV = 0
  PV = 0x556625c325e0 "\x01\x01\x02"\0
  CUR = 3
  LEN = 10
  COW_REFCNT = 0
  MAGIC = 0x556625c531d0
    MG_VIRTUAL = 0
    MG_TYPE = PERL_MAGIC_vstring(V)
    MG_LEN = 6
    MG_PTR = 0x556625c1bb60 "v1.1.2"
```
Note that the magic matches the PV.

I do think that the docs should indicate this is only useful when the PV and the magic are kept in sync (or someone is going to see unexpected results).

leonerd commented:
Having talked with @khwilliamson both on IRC and email, we are progressing with looking at how to add this to Devel::PPPort, but it seems the process is somewhat more involved than anticipated. It would be nice if that delay didn't hold up this PR, because it would ba shame for all API additions to be held hostage by the bus-factor of PPPort.

Leont commented:
Having thought more about it, I'm not sure `sv_vstring_set` should be public API (though it may well be fine as internal API). Nor am I sure it has the right shape in the first place. In its current form, you can easily set a vstring value that doesn't match with its string value, which seems wrong.

leonerd commented:
> Having thought more about it, I'm not sure `sv_vstring_set` should be public API (though it may well be fine as internal API). Nor am I sure it has the right shape in the first place. In its current form, you can easily set a vstring value that doesn't match with its string value, which seems wrong.

I'd be perfectly happy saying it shouldn't be public API. Unfortunately, Storable.xs needs to call it. So that's a bit out of the bag.

How about, instead of two separate steps of "make a plain SVt_PV", "add vstring magic to it", there was a single API function on "make a vstring, with these two stringy parts"? Effectively a dualstring var. ;)

Perhaps that's a more robust solution here?

tonycoz commented:
> How about, instead of two separate steps of "make a plain SVt_PV", "add vstring magic to it", there was a single API function on "make a vstring, with these two stringy parts"? Effectively a dualstring var. ;)

Are you looking at some particular use-case here?

If you have a x.y.z version then you can use new_version() to make a version SV out of it.

The only use I can see for this is for the deep serialisation that Storable does, Sereal, which I believe tries to be reasonably deep doesn't bother restoring version magic.

Leont commented:
> The only use I can see for this is for the deep serialisation that Storable does, Sereal, which I believe tries to be reasonably deep doesn't bother restoring version magic.

Exactly.

leonerd commented:
Honestly, the only "use case" is to get rid of as many lines of code that refer to `PERL_MAGIC_vstring` as possible. I'm looking for any way to achieve that. Storable.xs has an unfortunate rare cornercase in needing to create these, as well as inspect existing ones, so that makes it a bit difficult to just get rid of, without providing *some* kind of externally-visible API to let it do what it needs.

leonerd committed:
Add sv_vstring_get() API function and SvVSTRING() wrapper macro

This new function and wrapper macro mean that caller code does not have
to directly rely on (or be aware of) the `PERL_MAGIC_vstring` type. The
intent of the API now works independently of the current implemention as
magic.

leonerd commented:
As the `get` part of the API seemed relatively uncontentious, I have removed the `set` part for now to just this. Hopefully we can at least put this part in, and come back to look at the create or set side another time.

leonerd commented:
Actually now I look at this API, I wonder if it should be shaped to be similar to `SvPV`, as in:

```c
STRLEN vstr_len;
const char *vstr_pv = SvVSTRING(sv, vstr_len);
```

Leont commented:
> Actually now I look at this API, I wonder if it should be shaped to be similar to SvPV, as in:

That is also what I would have done.

leonerd commented:
Reviewers: PTAL. I've now added this as a `SvVSTRING` macro, and removed the `set` function for now. Hopefully this can go in in time for 5.41.10.
